### PR TITLE
feat: surface media usage in the task API and chat UI

### PR DESCRIPTION
**Pull request by OliverBryant**
Last of five changes splitting #997. Once this and the others land, #997
can be closed.

> **Depends on #1422, #1424, #1425 and #1457.** This branch contains their
> commits too, so the diff here shows all five. Review those first; once
> they merge this reduces to just the API + UI changes below.

The task detail endpoint now returns aggregated media rows, and the chat
token-usage popover renders them alongside LLM tokens.

`media_usage` carries one row per (model, unit, call_type, resolution)
from `aggregate_media_usage_by_model`. Deliberately **no cross-unit
quantity total** — summing images + seconds + characters produces a
number with no meaning — and **no `media_calls` companion field**, see
below. Locales cover all nine `MediaCallType` values and all five
`MediaUnit` values in both `en` and `zh`; I diffed the key sets against
the enums to confirm.

## Review feedback addressed from #997

- **The dead `media_calls` response field is dropped** rather than wired
  up. It had zero references in `frontend/src` while the component
  recomputed the same total from `media_usage`; a second server-side
  reduction is a duplicate that can drift from the rows it summarises.

- **Media rows are normalised at the fetch boundary.**
  `token_usage_details` is free-form legacy JSON, and both failure modes
  the review named were live: reducing over a `null` row throws on
  property access (blanking the component), and a string `quantity` like
  `"4"` passes the `> 0` check but then fails `Number.isFinite` in the
  formatter, so it silently rendered the "not yet measured" placeholder —
  reading as if the call had cost nothing. One normaliser now coerces
  every row to finite non-negative numbers and plain strings and drops
  non-objects. Self-review added arrays to that drop list, since
  `typeof [] === 'object'` and an array would otherwise normalise into a
  bogus all-zero row.

- **The test i18n mock is synced with the real locale files.** It carried
  a stale `unit.tokens` that no longer exists and never had `unit.texts`,
  which embedding rows use. One correction to the finding: the suite did
  already assert rendered unit labels (`"3 images"`, `"12.5 sec"`), so it
  was not fully uncatchable — but neither covered `texts`, which is
  exactly the key that was missing. Added a test asserting the rendered
  `texts` label so that specific drift is now caught.

Also added: null/non-object row survival and string-quantity rendering
tests. Both fail without the normaliser.

## Testing

CI. Frontend tests were not runnable in this worktree — vitest hangs
against a symlinked `node_modules` — and the Python suite does not run on
this machine at all (importing `xagent.core.model.chat` hangs, on `main`
too). The component changes were verified by review; `ruff`, `isort` and
`mypy` (pinned versions) are clean on the changed Python. I am relying on
CI's frontend jobs rather than claiming a local pass I did not get.

### CODE REVIEW

**Comment by OliverBryant** [comment]
Rebased onto latest `main` (`b2ddb5ad`). The diff is unchanged: 5 files, +601/-26.

**What was dropped.** This branch had been cut on top of the whole media-metering stack, so it carried 15 commits that do not belong to it:

- 9 commits of the **old** media usage primitives (`ffbecfed`..`f9eb91ee`) — superseded by #1527, which merged as `d74660c2`.
- #1424's 4 image-metering commits (`cd4db683`..`1b03632b`).
- #1425's 2 audio/video commits (`c2f09e4d`..`d1377994`).
- #1457's embedding/rerank commit (`26b1d612`).

Each of those is landing through its own PR, so only the top two commits — this PR's actual increment — were replayed onto `main`.

**No API migration was needed.** The increment's only reference to the primitives is `aggregate_media_usage_by_model`, which already merged with #1527 and is present on `main` (`src/xagent/core/model/chat/token_context.py`). Nothing here imports `standalone_usage` or any other symbol introduced by the sibling PRs, so this rebased directly onto `main` independently of #1425 and #1457.

One thing worth flagging so it does not read as an oversight: this branch does contain the string `MediaUnit`, but it is the **frontend TypeScript helper `formatMediaUnit`** in `TokenUsageDisplay.tsx`, not the Python `MediaUnit` enum. It needed no migration.

**Contract re-verified against current `main`**, since the enums moved under this branch:

- i18n `unit` map covers all 5 `MediaUnit` values (`images`, `seconds`, `characters`, `texts`, `requests`).
- i18n `mediaType` map covers all 9 `MediaCallType` values (`generate_image`, `edit_image`, `video`, `tts`, `asr`, `music`, `sound_effect`, `embedding`, `rerank`).
- The frontend `MediaUsage` interface matches the field set `aggregate_media_usage_by_model` actually returns.

**Drift.** `main` advanced 29 commits since this branch was cut; `en.ts`, `zh.ts` and `chat.py` all changed underneath it. The rebase replayed without conflicts and the resulting diff is additions-only in those three files, so no one else's work was reverted — in particular `main`'s `get_task` change from `get_latest_waiting_question` to `get_pending_interaction_question` is preserved alongside the new `media_usage` line.

**Checks run locally:** scoped `tsc --noEmit` over the four changed frontend files (clean), `ruff check` / `ruff format --check` and pinned `isort` on `chat.py` (clean), and pinned `mypy` — 118 pre-existing `--follow-imports=skip` findings, identical count on unmodified `main`, so no new ones. The vitest suite does not run in this environment; relying on CI for `TokenUsageDisplay.test.tsx`.

**Comment by OliverBryant** [comment]
One coverage caveat, filed separately as #1577 rather than fixed here to keep this PR a pure rebase.

`TokenUsageDisplay.test.tsx` is not executed by CI. The `frontend-build` job runs vitest through an explicit allowlist of launchers (`src/components/pages`, `src/components/kb`, `src/app`, five named `src/lib`/`src/i18n` files, plus the widget lane's per-file `include` list); `src/components/chat/` is in none of them. Confirmed on this PR's own `frontend-build` run: the string `TokenUsageDisplay` does not appear anywhere in the job log.

This is a pre-existing gap on `main`, not something this PR introduces — the file already carried 368 unexecuted lines before it. This PR takes it to 721. So the 401 added lines are typechecked (scoped `tsc --noEmit`, clean) and reviewable, but a green CI here does not mean they passed; it means they did not run. Worth weighing when reviewing the test file, since #1577 will run these for the first time.

File: frontend/src/components/chat/TokenUsageDisplay.tsx [hunk 1]
```diff
@@ -233,6 +318,83 @@ export function TokenUsageDisplay({ taskId, isRunning, className }: TokenUsageDi
...
+                    media.model_id,
+                    media.model_name,
+                    media.unit,
+                    media.call_type,
+                    media.resolution ?? "",
+                  ])}
+                >
+                  <span className="min-w-0" title={media.model_name || media.model_id}>
+                    <span className="block truncate font-medium">
+                      {media.model_name || media.model_id || t('chatPage.tokenUsage.unknownModel')}
```
Comment (rogercloud) [inline frontend/src/components/chat/TokenUsageDisplay.tsx:377]: **Minor:** The backend keeps media rows separate by `model_id`, but this renderer shows only `model_name` when both are present. Two configured IDs sharing a provider name are therefore visually indistinguishable. Please mirror the LLM row behavior by rendering a differing `model_id` as a secondary label/title and add a same-name/different-ID regression fixture.

File: frontend/src/components/chat/TokenUsageDisplay.tsx [hunk 2]
```diff
@@ -65,9 +114,18 @@ export function formatExactTokenCount(value: number, locale: Locale = 'en'): str
   return getFormatter(exactTokenFormatters, locale).format(normalizeTokenCount(value));
 }
 
+function formatMediaQuantity(value: number, locale: Locale = 'en'): string {
+  const safe = Number.isFinite(value) ? Math.max(0, value) : 0;
+  // Media quantities can be fractional (e.g. audio seconds); keep up to one
+  // decimal but drop trailing ".0" for whole counts like image counts.
+  return getFormatter(mediaQuantityFormatters, locale, {
```
Comment (rogercloud) [inline frontend/src/components/chat/TokenUsageDisplay.tsx:133]: **Minor:** `maximumFractionDigits: 1` rounds valid positive quantities below `0.05` to `0`, even though the raw value enters the measured branch. Please use adaptive precision or an explicit value such as `<0.1` for positive sub-threshold quantities, while preserving exact zero as unmeasured.

File: frontend/src/components/chat/TokenUsageDisplay.tsx [hunk 3]
```diff
@@ -128,6 +194,25 @@ export function TokenUsageDisplay({ taskId, isRunning, className }: TokenUsageDi
...
+  );
+  const formatMediaType = (callType: string) =>
+    callType
+      ? tDynamic(
+          `chatPage.tokenUsage.mediaType.${callType}`,
+          callType,
+        )
+      : '-';
+  const formatMediaUnit = (unit: string) =>
+    unit ? tDynamic(`chatPage.tokenUsage.unit.${unit}`, unit) : '';
```
Comment (rogercloud) [inline frontend/src/components/chat/TokenUsageDisplay.tsx]: **Minor:** The fixed English unit labels produce `1 images`, `1 chars`, `1 requests`, and `1 texts` for valid singular quantities (`sec` is fine as an abbreviation). Please use quantity-aware singular/other translations and add singular-row assertions.

File: src/xagent/web/api/chat.py [hunk 4]
```diff
@@ -4589,6 +4593,12 @@ def _get_task_sync() -> Dict[str, Any]:
                     entry["cache_write_input_tokens"] for entry in model_usage
                 ),
                 "model_usage": model_usage,
+                # No media_calls companion: the client derives its own count
+                # from these rows, and a second server-side reduction would be
+                # a duplicate that can drift. Deliberately no cross-unit
+                # quantity total either — summing images + seconds + characters
+                # produces a number with no meaning.
+                "media_usage": media_usage,
```
Comment (rogercloud) [inline src/xagent/web/api/chat.py:4601]: **Minor:** The new `media_usage` response field is not covered at the real task-detail serialization/cache seam. Please extend the existing GET route test with representative media details, exact response assertions, and preferably a cache-hit assertion.

File: frontend/src/components/chat/TokenUsageDisplay.test.tsx [hunk 5]
```diff
@@ -265,6 +289,335 @@ describe("TokenUsageDisplay", () => {
...
+    )
+
+    render(<TokenUsageDisplay taskId={31} isRunning={false} />)
+
+    fireEvent.click(await screen.findByRole("button", { name: /1 media call/ }))
+
+    expect(await screen.findByText("32 texts")).toBeInTheDocument()
+  })
+
+  it("survives null and non-object media rows", async () => {
```
Comment (rogercloud) [inline frontend/src/components/chat/TokenUsageDisplay.test.tsx:619]: **Minor:** This malformed-row test omits arrays, so removing the production `Array.isArray` guard could add an `Unknown` zero row while the current count/quantity assertions still pass. Please include `[]` and assert no extra rendered invalid row.

File: frontend/src/components/chat/TokenUsageDisplay.tsx [hunk 6]
```diff
@@ -233,6 +318,83 @@ export function TokenUsageDisplay({ taskId, isRunning, className }: TokenUsageDi
...
+                  </span>
+                  <span className="min-w-0 truncate" title={formatMediaType(media.call_type)}>
+                    {formatMediaType(media.call_type)}
+                  </span>
+                  <span className="text-right tabular-nums">
+                    {/* quantity 0 means the call happened but the provider
+                        reported no measurable size (e.g. an async video with
+                        no duration yet) — say so rather than showing "0 sec",
+                        which reads as "this cost nothing". */}
+                    {media.quantity > 0
```
Comment (rogercloud) [inline frontend/src/components/chat/TokenUsageDisplay.tsx:403]: **Minor:** No frontend fixture reaches the zero/unmeasured branch, and the test mock omits `chatPage.tokenUsage.unmeasured`. Please add that mock key plus a zero-quantity row with `calls > 0` and assert the localized `not yet measured` text.

File: frontend/src/components/chat/TokenUsageDisplay.test.tsx [hunk 7]
```diff
@@ -265,6 +289,335 @@ describe("TokenUsageDisplay", () => {
...
+        }),
+        { status: 200, headers: { "Content-Type": "application/json" } },
+      ),
+    )
+
+    render(<TokenUsageDisplay taskId={25} isRunning={false} />)
+
+    fireEvent.click(await screen.findByRole("button", { name: /2 media calls/ }))
+    // Real Gemini image tokens are surfaced, not silently dropped.
+    expect(await screen.findByText(/1\.12k/)).toBeInTheDocument()
```
Comment (rogercloud) [inline frontend/src/components/chat/TokenUsageDisplay.test.tsx:865]: **Minor:** The provider-token checks match only `1.12k` and `40~`, so a missing or wrong `tokensShort` label can pass. Please add the mock key and assert complete strings such as `1.12k tokens` and `40~ tokens`.

File: frontend/src/components/chat/TokenUsageDisplay.test.tsx [hunk 8]
```diff
@@ -265,6 +289,335 @@ describe("TokenUsageDisplay", () => {
...
+              call_type: "generate_image",
+              resolution: "1K",
+              quantity: 3,
+              calls: 2,
+              tokens: 0,
+            },
+            {
+              model_id: "tts-1",
+              model_name: "elevenlabs-tts",
+              unit: "seconds",
```
Comment (rogercloud) [inline frontend/src/components/chat/TokenUsageDisplay.test.tsx]: **Minor:** This fixture pairs `call_type: "tts"` with `unit: "seconds"`, but the producer contract maps TTS to `characters`, so the real TTS path is untested. Please use `unit: "characters"` and assert `chars`; use video or ASR for seconds.

**Review by rogercloud — APPROVED** [review]
## Summary

PR #1463 completes the read-side media-usage slice by projecting persisted media detail rows through `GET /api/chat/task/{id}` and rendering them in the chat usage popover. It reuses the existing aggregation, task-detail cache, polling lifecycle, and localization machinery while keeping heterogeneous units separate and preserving additive API compatibility.
Blocking: no — recommended event: APPROVE

## Update summary

The update consists of commits `026a3a6f` and `bb428d95`. `026a3a6f` adds the task-detail API aggregation/serialization, media popover, English and Chinese labels, and frontend coverage; `bb428d95` fixes the TypeScript annotation/noImplicitAny issue. The rebase context in [conversation comment 5365048088](https://github.com/xorbitsai/xagent/pull/1463#issuecomment-5365048088) explains that the superseded producer stack was dropped/replayed, so this review evaluates the current five-file API/UI slice.

## Round 0 design verdict

**Verdict: acceptable-with-reservations.** The direction is sound: the PR uses the existing persisted `TokenUsage.details` stream and media aggregator, adds one additive task-detail field, and uses the established popover, polling, and dynamic-i18n patterns. The mixed measured/unmeasured aggregate-state concern is the exact tracked root in [#1495](https://github.com/xorbitsai/xagent/issues/1495), and the cumulative JSON persistence/read scaling boundary is the exact tracked root in [#1466](https://github.com/xorbitsai/xagent/issues/1466); both are dropped context rather than new findings for this PR. The remaining design reservation is the missing real task-detail serialization/cache contract test, reported below as `NEW-API-SEAM-TEST`; the independently raised API-shape ownership concern was dropped because the backend already follows the repository's dynamic `Dict[str, Any]` convention and the frontend normalizer is an appropriate network/legacy-data boundary.

## Prior-findings checklist (body-only and inline roots)

| Canonical root | Status | Evidence and disposition |
| --- | --- | --- |
| `TokenUsage.merge` self-merge doubling | **DROPPED — pre-existing/out of scope** | The body-only review finding [4959106301](https://github.com/xorbitsai/xagent/pull/1463#pullrequestreview-4959106301) and its duplicate inline root [3802478157](https://github.com/xorbitsai/xagent/pull/1463#discussion_r3802478157) both concern `src/xagent/core/model/chat/token_context.py`, which is byte-identical between the supplied base and head and is not in this PR's diff. The current in-tree caller constructs a fresh target before merging, and [conversation comment 5365048088](https://github.com/xorbitsai/xagent/pull/1463#issuecomment-5365048088) explains the superseded stack/rebase context. This is not a `FIXED` claim and is not re-reported. |
| Frontend chat test CI execution gap | **DROPPED — tracked pre-existing** | [Conversation comment 5365089129](https://github.com/xorbitsai/xagent/pull/1463#issuecomment-5365089129) identifies the existing explicit-launcher gap for `frontend/src/components/chat/TokenUsageDisplay.test.tsx`; the workflow, package scripts, and manifest are unchanged by this PR. The concern is tracked in [#1577](https://github.com/xorbitsai/xagent/issues/1577), so it is not re-reported as a PR finding. |

The mixed measured/unmeasured aggregate concern is likewise dropped as tracked context only under [#1495](https://github.com/xorbitsai/xagent/issues/1495), and the cumulative details persistence/scaling concern is dropped as tracked context only under [#1466](https://github.com/xorbitsai/xagent/issues/1466). Neither is an active finding below.

## Review-criteria coverage

- **Code quality and local API design:** The implementation reuses existing abstractions without adding a parallel counter, endpoint, or speculative schema framework. The model-identity presentation issue is reported below; the broader contract-ownership candidate was independently dropped as non-actionable.
- **Code correctness and edge cases:** The review checked malformed/legacy JSON, numeric coercion, missing fields, unknown values, zero quantities, small positive quantities, unit/call-type identity, cache behavior, polling cleanup, authorization, and escaping. The confirmed display edge cases are listed below; no new resource-lifecycle, concurrency, error-propagation, security, or compatibility blocker was found.
- **Test quality and coverage:** Existing aggregation, locale-parity, and lower-level persistence tests were considered. The missing API seam assertion and four distinct frontend test/fixture gaps (arrays, zero/unmeasured, provider-token labels, and TTS units) are listed below.
- **Design and architecture:** The read-model/UI direction fits the existing `TokenUsage.details` pipeline and correctly avoids cross-unit totals. The two pre-existing architectural boundaries are tracked in #1495 and #1466 rather than duplicated here.
- **Documentation and localization:** Both shipped locale trees contain the current media unit/type keys and preserve unknown-value fallback behavior. No separate documentation issue was found because this internal legacy endpoint has no existing field-level response documentation convention.

## New findings (all minor, non-blocking)

1. **`frontend/src/components/chat/TokenUsageDisplay.tsx:357` — minor — media rows hide meaningful model identity.** The backend groups by `model_id or model_name` and preserves both fields, but this renderer shows only `model_name` when it is present. Two distinct configured IDs sharing a provider-facing name therefore produce visually indistinguishable rows. **Suggestion:** mirror the existing LLM-row behavior: when `model_id` and `model_name` are both non-empty and differ, render `model_id` as a secondary label/title, and add a same-name/different-ID regression fixture.

2. **`frontend/src/components/chat/TokenUsageDisplay.tsx:121` — minor — small positive quantities can display as zero.** `Intl.NumberFormat` with `maximumFractionDigits: 1` rounds a valid `0 < quantity < 0.05` to `0`, while the renderer treats the raw value as measured. **Suggestion:** use adaptive precision or an explicit representation such as `<0.1` for positive sub-threshold values, while keeping exact zero mapped to the unmeasured state.

3. **`frontend/src/components/chat/TokenUsageDisplay.tsx:214` — minor — fixed plural unit labels are grammatically wrong for quantity one.** The new English labels render values such as `1 images`, `1 chars`, `1 requests`, and `1 texts`; `sec` is an acceptable invariant abbreviation. **Suggestion:** add quantity-aware singular/other translations (for example, `1 image`, `1 char`, `1 request`, and `1 text`) and assert the singular row labels.

4. **`src/xagent/web/api/chat.py:4601` — minor — the new task-detail serialization seam lacks a route/cache contract test.** Aggregator tests and mocked frontend tests can pass if the production GET handler omits, renames, or fails to cache `media_usage`. **Suggestion:** extend the existing real `GET /api/chat/task/{id}` test with representative media details and exact `media_usage` assertions, including zero quantity, calls, resolution, provider tokens, and `tokens_estimated`; preferably assert the cache-hit response too.

5. **`frontend/src/components/chat/TokenUsageDisplay.test.tsx:387` — minor — the malformed-row test does not cover arrays or prove invalid rows are absent.** The production normalizer explicitly rejects arrays, but removing that guard would create an extra zero/unknown row while the current count and quantity assertions could still pass. **Suggestion:** add an array-shaped row such as `[]` and assert exactly the expected rendered row (or the absence of an `Unknown`/unmeasured junk row), retaining the null/string/number cases.

6. **`frontend/src/components/chat/TokenUsageDisplay.tsx:373` — minor — the zero/unmeasured rendering branch has no frontend fixture.** All current media quantities are positive and the test translation mock omits `chatPage.tokenUsage.unmeasured`, so a regression to `0 sec` or a missing translation can pass. **Suggestion:** add the mock key and a zero-quantity media row with `calls > 0`, then assert the localized `not yet measured` text rather than numeric zero.

7. **`frontend/src/components/chat/TokenUsageDisplay.test.tsx:615` — minor — provider-token tests assert numbers but not the token label.** The checks for `1.12k` and `40~` would pass if `tokensShort` were missing, mistyped, or rendered with the wrong translation key. **Suggestion:** add `chatPage.tokenUsage.tokensShort` to the mock and assert the complete strings, such as `1.12k tokens` and `40~ tokens`.

8. **`frontend/src/components/chat/TokenUsageDisplay.test.tsx:325` — minor — the TTS fixture uses an impossible unit pairing.** The fixture sets `call_type: "tts"` with `unit: "seconds"`, but the producer contract maps TTS to `characters`, so it never tests the actual TTS path. **Suggestion:** change the fixture to `unit: "characters"` and assert the `chars` label; use video or ASR for a seconds-formatting fixture.

## Simplification Lens status

The dedicated Simplification Lens scan was unavailable because it returned `usage_limit_reached`; no simplification findings are asserted.

## Validation context

CI preflight was green with no failed or pending checks. No local tests, builds, linters, formatters, or dependency installation were run; local validation was intentionally not run under the review policy.

## Blocking status & recommended decision

**Blocking: no.** There are no confirmed major or critical findings. All eight confirmed findings above are minor, and the prior self-merge and frontend CI-gap roots were dropped as pre-existing/out-of-scope or tracked.

**Recommended event: APPROVE**

File: frontend/src/components/chat/TokenUsageDisplay.tsx (hunk 1 above)
Comment (OliverBryant) [reply frontend/src/components/chat/TokenUsageDisplay.tsx:377]: Fixed in c21e40ce. The media row now applies the same secondary-identity rule as the LLM rows: when `model_id` and `model_name` are both non-empty and differ, the id renders as a secondary label. Added two fixtures — two rows sharing `stable-diffusion-xl` with ids `sd-primary`/`sd-fallback` (both ids asserted present, the shared name asserted exactly twice), and an `id === name` row asserting the name is not duplicated.

One deliberate omission: I did not mirror the LLM rows' `!model_id && model_name` → "Unattributed" badge. That badge belongs to the attributed-vs-unattributed model count in the LLM popover header, which the media popover has no counterpart for (it shows a plain call count). Adding the badge without that surrounding concept would label a state the media UI never explains. Happy to add it if you want the concept extended to media rows.

File: frontend/src/components/chat/TokenUsageDisplay.tsx (hunk 2 above)
Comment (OliverBryant) [reply frontend/src/components/chat/TokenUsageDisplay.tsx:133]: Fixed in c21e40ce. `formatMediaQuantity` now returns a localized `<0.1` for `0 < quantity < 0.05` rather than letting it round to `0`. Exact zero never reaches the formatter — the call site routes it to the unmeasured branch — so the two states stay distinct. Added a 0.02s ASR fixture asserting `<0.1 sec` and asserting neither `0 sec` nor the unmeasured text appears; verified the fix is load-bearing by reverting the guard, which yields `0 sec`.

File: frontend/src/components/chat/TokenUsageDisplay.tsx (hunk 3 above)
Comment (OliverBryant) [reply frontend/src/components/chat/TokenUsageDisplay.tsx]: Fixed in c21e40ce. Unit labels are now `one`/`other` pairs in both locale trees and selected by the raw quantity, so a single item reads `1 image`, `1 char`, `1 request`, `1 text`. `sec` is intentionally identical in both branches as an invariant abbreviation, and zh carries the same text in both since Chinese has no plural distinction — the locale trees stay structurally identical, which `translations.test.ts` enforces.

Plural selection uses the raw quantity rather than the formatted string, so the new sub-threshold `<0.1` case still reads as plural. Added a fixture asserting the three singular labels and asserting the plural forms are absent; existing plural fixtures (`3 images`, `32 texts`, `480 chars`) still pass.

File: src/xagent/web/api/chat.py (hunk 4 above)
Comment (OliverBryant) [reply src/xagent/web/api/chat.py:4601]: Fixed in c21e40ce. Added two tests to `tests/web/test_chat_task_model_ids.py`, alongside the existing `model_usage` route test:

- `test_get_task_returns_media_usage_grouped_by_model_and_unit` asserts the exact `media_usage` payload from a real `GET /api/chat/task/{id}`, covering two entries collapsing on the same (model, unit, call_type, resolution) key with summed quantity/calls, `tokens_estimated: true` propagating from one estimated entry in a mixed group, a zero-quantity row surviving with `resolution: ""` and `provider_tokens: 0`, and an LLM entry staying out of `media_usage` (with `model_usage` asserted to still hold it).
- `test_web_task_detail_cache_serves_media_usage_on_cache_hit` asserts the field is identical on the cache-hit read, since the detail response is cached wholesale and the popover polls that path.

Both expected payloads were checked against the live aggregator output rather than hand-written.

File: frontend/src/components/chat/TokenUsageDisplay.test.tsx (hunk 5 above)
Comment (OliverBryant) [reply frontend/src/components/chat/TokenUsageDisplay.test.tsx:619]: Fixed in c21e40ce. Added `[]` to the malformed-row list and added absence assertions (`Unknown model` and the unmeasured text). Your read was exactly right and worth spelling out: I removed the `Array.isArray` guard to check, and the pre-existing `1 media call` and `2 images` assertions both still passed — the leaked array normalises to an all-zero row, so it adds a row without changing the call total. Only the new absence assertions fail without the guard.

File: frontend/src/components/chat/TokenUsageDisplay.tsx (hunk 6 above)
Comment (OliverBryant) [reply frontend/src/components/chat/TokenUsageDisplay.tsx:403]: Fixed in c21e40ce. Added `chatPage.tokenUsage.unmeasured` to the test locale mock and a fixture with `quantity: 0, calls: 2` (a video row), asserting the localized `not yet measured` text, that the row is still counted in the `2 media calls` trigger, and that `0 sec` does not appear. Worth noting why the mock key mattered: the mock's `tDynamic` falls back to the supplied fallback on a miss, so without the key registered the assertion could have passed against a fallback string rather than a real translation.

File: frontend/src/components/chat/TokenUsageDisplay.test.tsx (hunk 7 above)
Comment (OliverBryant) [reply frontend/src/components/chat/TokenUsageDisplay.test.tsx:865]: Fixed in c21e40ce. Added `chatPage.tokenUsage.tokensShort` to the mock and now assert the complete strings `1.12k tokens` and `40~ tokens`. The count and label are separate JSX children, so the assertion normalises whitespace and anchors the full text rather than matching a substring — a missing, misnamed, or wrongly-resolved label now fails instead of passing on the number alone.

File: frontend/src/components/chat/TokenUsageDisplay.test.tsx (hunk 8 above)
Comment (OliverBryant) [reply frontend/src/components/chat/TokenUsageDisplay.test.tsx]: Fixed in c21e40ce. Changed the fixture to `unit: "characters"` with `quantity: 480` and it now asserts `480 chars`; added a separate `whisper-large` ASR row at `12.5` for the seconds path, asserting `12.5 sec`. Confirmed against `MediaCallType` on main — `TTS = ("tts", MediaUnit.CHARACTERS)` and the unit is derived from the modality, so the old tts/seconds pair was unrepresentable upstream and the assertion pinned a state the producer can never emit. The call-count assertion moved from 3 to 4 for the added row.

File: src/xagent/core/model/chat/token_context.py [hunk 9]
```diff
@@ -51,62 +135,132 @@ def add_input_tokens(
...
     def merge(self, other: "TokenUsage") -> None:
         """Merge another TokenUsage into this one."""
-        self.input_tokens += other.input_tokens
-        self.output_tokens += other.output_tokens
-        self.llm_calls += other.llm_calls
-        self.tool_calls += other.tool_calls
-        self.details.extend(other.details)
+        # Snapshot ``other`` under its own lock and release it before taking
+        # ours: holding both at once would deadlock a concurrent ``b.merge(a)``.
+        with other._lock:
```
[the remark this answers is not in the document]
Comment (OliverBryant) [reply src/xagent/core/model/chat/token_context.py]: Not applicable to this PR. `src/xagent/core/model/chat/token_context.py` is not in this PR's diff — the branch was rebased to drop the superseded media-primitives stack it had been cut on top of (see the rebase note in this thread's conversation), and this PR now touches only the task-detail API field and the chat UI. This thread is marked outdated for that reason.

The `merge` self-merge concern belongs to the primitives that landed via #1527. On current main the in-tree caller constructs a fresh target before merging, so the doubling path is not reachable there today; if a guard is still wanted it should go in against that file directly rather than through this read-side PR.
